### ISSUE TITLE
Expand selection to the left/right of cursor

### ISSUE THREAD

**Issue by casperc** [cursiveclojure/cursive#997]
I am currently using Counterclockwise for editing clojure, but I would be interested in switching.

One feature for paredit that Counterclockwise has is to select the form to the left or right of the curser, which is very useful.

So you have the following form, with the cursor at "|" and you press the key to select the form to the left:

`((juxt :one :two)| m)`

In this case, the form `(juxt :one :two)` is selected. If the curser is at the other side and you press the key to select the form to the right:

`(|(juxt :one :two) m)`

you get the same form.

It (for me) creates are more intuitive way to move around and select the code and comparing to using expand selection, it saves some keypresses. It also provides the option to keep expanding the selection in that direction to select several forms or symbols next to each other.

I am aware that expand selection does some of this, however it has the counterintuitive behaviour that if you have the cursor at the following position:

`(do-something (one)|)`

it will select the entire form, not just the `(one)` form, which is even marked by the par-matching at that position. If it would select the matched form, I would be more satisfied with the behaviour. However at present (as far as I am aware), in order to select the `(one)` form, I need to move into the form and press "expand selection" twice, which I think is quite a few keypresses for a common use case.

**Comment by cursiveclojure** [comment]
Yes, this functionality is very nice. There's been some [discussion](https://cursiveclojure.com/archive/840.html) on the mailing list about how best to handle this. I actually like the CCW model the best, I think, but there's an issue with it in IntelliJ. In CCW, Alt-up works like Expand Selection, and Alt-left/right expand it laterally. Critically, Alt-down shrinks whatever the last expansion was, either a normal expansion or a lateral one. Unfortunately that's difficult in IntelliJ, since Expand/Shrink selection are built into IntelliJ and work for all languages - if I change how those work it will be very confusing for experienced IntelliJ users. But I'm going to look to see if I can customise that for Clojure without breaking anything else.

**Comment by cursiveclojure** [comment]
BTW I've fixed the list expansion case you mention for the next build, I agree that that was counterintuitive. So now, expanding the selection at the end of a list will select that list.

**Comment by casperc** [comment]
Cool, I'll look out for when it is out.

**Comment by mbossenbroek** [comment]
+1000 for this feature. Coming from CCW I'm definitely missing the left/right expansion.

Would it be possible to add brand new keybindings for these behaviors? That way you could leave the existing functionality as-is and those who want the CCW style could re-bind as needed.

Instead of expand/shrink, it could be expand/expand-left/expand-right/previous-expand (or undo-expand).

**Comment by cursive-ide** [comment]
@mbossenbroek Yes, that's pretty much what I'm planning. After some discussion with JetBrains, I came up with a strategy to keep the existing actions and add the left/right expansions too.

**Comment by ataggart** [comment]
Is there any news on this feature? It's the one thing I miss from CCW.

**Comment by imrekoszo** [comment]
@cursive-ide I think the link pointing to the discussion in one of your earlier comments is broken. Here's what I was able to find searching the google group: https://groups.google.com/d/topic/cursive/yLGvyk3kl8w/discussion

Is there any hope this will get closer to being addressed?

There are a few issues I've recently became aware of that appear to be related: #1022, #1705, #2258  

I think if extending the selection sideways becomes possible that could serve as a solution to the 'select key-value-pair' problem. (Use extend selection until you have either the key or the value selected then extend to the left or right). It could perhaps also be a more general solution, as semantic pairs can occur not just in `let` and `{}` forms but many other places too, including `case`, `cond`, custom macros etc.

Moreover, in other contexts, like the body of `clojure.test/are`, there is a semantic group between the single form and entire sexp level, just as pairs in a map or let, but the number of forms that make up that semantic group depends on surrounding code. In the following example it would make sense to be able to select groups of 3:
```clojure
(deftest +-test
  (are [a b result] (= result (+ a b))
    0 0 0
    0 1 1
    1 0 1
    1 1 2))
```

**Comment by imrekoszo** [comment]
I was also going to reply to the thread in the google group but I think it is probably better if I just do it here.

> I’m not sure how much this adds over Toggle Sticky Selection and then moving using the normal commands.

While sticky selection works okay, I have 2 frustrations with it:

**One**: My existing selection gets canceled.

Say I'm in the following position, with | indicating the cursor
```clojure
[{:foo {:aaa 1
        :b|bb 2
        :ccc 3}}
 {:qux {:asd :qwe}}]
```
and want to end up with the following selection, indicated by _s:
```clojure
[{:foo {:aaa 1
        _:bbb 2
        :ccc 3_}}
 {:qux {:asd :qwe}}]
```
What would be intuitive is to
1. extend selection until I reached the level I want to start selecting siblings
2. select siblings

But with sticky selection what happens is once I reach
```clojure
[{:foo {:aaa 1
       _ :b|bb_ 2
        :ccc 3}}
 {:qux {:asd :qwe}}]
```
and turn on sticky and move to the right, I end up with
```clojure
[{:foo {:aaa 1
        :b_bb|_ 2
        :ccc 3}}
 {:qux {:asd :qwe}}]
```
So currently, before starting the sticky selection I have to first place the cursor before the first (or after the last) sibling I want to include in the selection. If the original cursor position was several levels down, this can be frustrating.

**Two**: it is too easy to end up with an unbalanced selection using sticky + move forward/backward

So let's imagine I'm at
```clojure
[{:foo {:aaa 1
        |:bbb 2
        :ccc 3}}
 {:qux {:asd :qwe}}]
```
and again I want to select
```clojure
[{:foo {:aaa 1
        _:bbb 2
        :ccc 3_}}
 {:qux {:asd :qwe}}]
```
but imagine that the map is a lot larger, perhaps it's several screens. If I turn on sticky and hold down the `Move Forward` command, I can easily end up with
```clojure
[{:foo {:aaa 1
        _:bbb 2
        :ccc 3}}
 {:qux {:asd :qwe|_}}]
```
which is most certainly not what I wanted.

> unbalance the selection

**Ideally** as I imagine it now, a structural extend to sides selection would never end up with an unbalanced selection, so if one wanted to cross the boundary of the current sexp they'd have to extend up first, to select the entire sexp, then they'd be able to extend further to the sides.

So in addition to the current Extend selection and Shrink selection I'd probably add 4 commands to `Extend to`/`Shrink from` left/right.

I'd also disable structural side-selection and all structural editing commands entirely if the current selection is unbalanced.
